src/traverse/iter.rs

```diff
@@ -0,0 +1,200 @@
+use super::*;
+use crate::Documentation;
+
+use std::fs;
+
+use log::{trace, warn};
+
+use std::path::{Path, PathBuf};
+
+use anyhow::{anyhow, Error, Result};
+
+/// An iterator traversing module hierarchies yielding paths
+#[derive(Debug, Clone)]
+pub struct TraverseModulesIter {
+    /// state for enqueuing child files and the depth at which they are found
+    queue: VecDeque<(PathBuf, usize)>,
+    /// zero limits to the provided path, if it is a directory, all children are collected
+    max_depth: usize,
+}
+
+impl Default for TraverseModulesIter {
+    fn default() -> Self {
+        Self {
+            max_depth: usize::MAX,
+            queue: VecDeque::with_capacity(128),
+        }
+    }
+}
+
+impl TraverseModulesIter {
+    fn add_initial_path<P>(&mut self, path: P, level: usize) -> Result<()>
+    where
+        P: AsRef<Path>,
+    {
+        let path = path.as_ref();
+        let path = path.canonicalize().unwrap();
```

Comment:
Canonicalize should only fail in https://doc.rust-lang.org/src/std/fs.rs.html#1847-1848 cases, so yes, this should be fixed at some point, but it does not hurt right now, since we check the metadata before we ever feed the path to this function, which guarantees it's either a file or directory.